Rubberduck.Refactorings/Abstract/CodeOnlyRefactoringActionBase.cs

```diff
@@ -0,0 +1,40 @@
+﻿using Rubberduck.Parsing.Rewriter;
+using Rubberduck.Parsing.VBA.Parsing;
+using Rubberduck.Refactorings.Exceptions;
+
+namespace Rubberduck.Refactorings
+{
+    public abstract class CodeOnlyRefactoringActionBase<TModel> : ICodeOnlyRefactoringAction<TModel>
+        where TModel : class, IRefactoringModel
+    {
+        private readonly IRewritingManager _rewritingManager;
+
+        protected CodeOnlyRefactoringActionBase(IRewritingManager rewritingManager)
+        {
+            _rewritingManager = rewritingManager;
+        }
+
+        public abstract void Refactor(TModel model, IRewriteSession rewriteSession);
```

Comment:
The `Refactor` signature generated some thoughts.  Taking `RenameRefactoring` as an example, what `RenameRefactoringAction` _really_ wants is a Declaration/String pair (target/new name).  So, being relieved of participating in the MVVM transactions, `RenameRefactoringAction` has the opportunity to ignore that a TModel even exists and accept a collection of Declaration/String pairs as the input.  To do this would basically require an `IRenameRefactoringAction` that extends the `ICodeOnlyRefactoringAction` interface.

Thought/question is this...It would seem preferable for any `TRefactoringAction` to extend the `ICodeOnlyRefactoringAction` interface in order to support additional input sets.  Of course, behind the interface, the alternative input set would probably be loaded into a new `TModel` instance by the `TRefactoringAction` - but doing so is now optional - and all TModel interactions are encapsulated behind the interface.  Thoughts?

Reply:
First of all, the `TModel` does not necessarily have anything to do with any presentation to the user, it is just a container for the configuration the refactoring requires. The refactoring actions are specifically generic, as well as the interfaces, to allow for different kinds of input.

Btw, the rename refactoring really needs more information than just the old declaration and the new name. You also need to know whether related declarations need to be renames, e.g. interface implementations. If you look at the implementation of `RenameModel`, you can see that there really is not much apart from the necessary.

Moreover, note that the rename refactoring is now basically implemented by two refactoring actions, one for components and projects, which really cannot be an `ICodeOnlyRefactoringAction`, and one for all otherdseclarations, which is a, `ICodeOnlyRefactoringAction`.

Please note that the `ICodeOnlyRefactoringAction<TModel>` interface is only meant for refactoring actions without side-effects apart from manipulating the rewrite session. Everything else would basically invalidate the known state for the caller.